### DISCUSSION TITLE
Automatic Python Version Constraining

### DISCUSSION THREAD

**Discussion by mitsuhiko** [astral-sh/rye#64] [Ideas]
While fixing #63 I realized that I did not actually think much about Python version constraining before. With the changes from d32c7a7aee89db85b9abc9ad9d3975a586d66354 the internal resolver _when adding a new package_ now automatically assumes that the supported Python version in the `pyproject.toml` sets the lower bound.

This now means the following. When you declare something like this in your pyproject.toml:

```toml
requires-python = ">= 3.7"
```

Adding a new dependency takes that into account for as long as the dependency declares that. Which in turn means that for instance `rye add django` will not pick an incompatible Python version:

```
$ rye add django
Added django~=3.2.18
```

If you had set `3.8` as lower boundary, it would pick a much near django:

```
$ rye add django
Added django~=4.2
```

This now also means that `rye show` spits out a `min python` in addition to the `pinned python` which is the Python that is actually used for the virtualenv:

```
$ rye show
project: hello-world
path: /Users/mitsuhiko/hello-world
venv: /Users/mitsuhiko/hello-world/.venv
min python: 3.7
pinned python: cpython@3.11.1
workspace: /Users/mitsuhiko/hello-world
  members:
    hello-world (./)
    hello-support (./hello-support)
```

Does this make sense or is this confusing?

**Comment by cnpryer** [comment]
> Does this make sense or is this confusing?

Makes sense to me.
